### PR TITLE
[skillshare:course] Add new extractor

### PR DESCRIPTION
**Pull request by Oshawk**
Added new extractor for skillshare.com classes (you can't link to a specific video).

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added support for [skillshare.com](https://skillshare.com/) classes such as [this](https://www.skillshare.com/classes/Digital-Illustration-Learn-to-Use-Procreate/971736290) using APIs calls extracted from the mobile app.
Tested with both premium and free courses and premium and free users. Works like any other extractor, requires username and password.
Individual video support not added as I can find no way of producing a URL for a specific video.

### CODE REVIEW

File: youtube_dl/extractor/skillshare.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import json
+
+from .common import InfoExtractor
+from ..compat import compat_str
+from ..utils import ExtractorError
+from ..utils import int_or_none
+
+
+class SkillshareBaseIE(InfoExtractor):
+    _NETRC_MACHINE = 'udemy'
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Wrong id.

File: youtube_dl/extractor/skillshare.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    _TN_RE = r"uploads/video/thumbnails/[0-9a-f]+/(?P<width>[0-9]+)-(?P<height>[0-9]+)"
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Single quotes.

File: youtube_dl/extractor/skillshare.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    _LOGIN_URL = "https://api.skillshare.com/login"
+    _VIDEO_URL = "https://api.skillshare.com/sessions/%s/download"
+
+    def _real_initialize(self):
+        self._login()
+
+    def _login(self):
+        username, password = self._get_login_info()
+        if username is None or password is None:
+            self.raise_login_required("An email and password is needed to download any video (even non-premium ones)")
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Must allow cookie authentication.

File: youtube_dl/extractor/skillshare.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+from datetime import datetime
+import json
+
+from .common import InfoExtractor
+from ..compat import compat_str
+from ..utils import ExtractorError
+from ..utils import int_or_none
+
+
+class SkillshareBaseIE(InfoExtractor):
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: There is no point in base class.

File: youtube_dl/extractor/skillshare.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        "url": "https://www.skillshare.com/classes/Blender-3D-Fire-Smoke-Simulation-Guide/1850126092",
+        "only_matching": True
+    }
+
+    def _real_extract(self, url):
+        # Technically the SKU, not ID but the SKU is a more universal identifier.
+        class_id = self._match_id(url)
+        class_json = self._download_json(self._CLASS_URL % class_id,
+                                         None,
+                                         note="Getting class details",
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: `Downloading class JSON`.

File: youtube_dl/extractor/skillshare.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            for lesson_thumbnail_url in lesson_thumbnail_urls:
+                lesson_thumbnails_json.append({
+                    "url": lesson_thumbnail_url,
+                    "width": int_or_none(self._search_regex(self._TN_RE, lesson_thumbnail_url, "width", fatal=False)),
+                    "height": int_or_none(self._search_regex(self._TN_RE, lesson_thumbnail_url, "height", fatal=False)),
+                })
+
+            try:
+                lesson_timestamp_dt = datetime.strptime(lesson_json.get("create_time", ""), "%Y-%m-%d %H:%M:%S")
+                lesson_timestamp = int(lesson_timestamp_dt.strftime("%s"))
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: unified_timestamp

File: youtube_dl/extractor/skillshare.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            except ValueError:
+                lesson_timestamp = None
+
+            videos.append({
+                "id": str(lesson_json["id"]),
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: compat_str

File: youtube_dl/extractor/skillshare.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                "title": lesson_json.get("title"),
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Title is mandatory.

File: youtube_dl/extractor/skillshare.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                "url": self._VIDEO_URL % str(lesson_json["id"]),
+                "ext": "mp4",
+                "thumbnails": lesson_thumbnails_json,
+                "uploader": class_json["_embedded"].get("teacher", {}).get("full_name"),
+                "creator": class_json["_embedded"].get("teacher", {}).get("full_name"),
+                "timestamp": lesson_timestamp,
+                "uploader_id": str(class_json["_embedded"].get("teacher", {}).get("username", 0)),
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Lack of data is denoted by `None` not `int`.

File: youtube_dl/extractor/skillshare.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                lesson_timestamp = None
+
+            videos.append({
+                "id": str(lesson_json["id"]),
+                "title": lesson_json.get("title"),
+                "url": self._VIDEO_URL % str(lesson_json["id"]),
+                "ext": "mp4",
+                "thumbnails": lesson_thumbnails_json,
+                "uploader": class_json["_embedded"].get("teacher", {}).get("full_name"),
+                "creator": class_json["_embedded"].get("teacher", {}).get("full_name"),
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: `try_get`.

File: youtube_dl/extractor/skillshare.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                "timestamp": lesson_timestamp,
+                "uploader_id": str(class_json["_embedded"].get("teacher", {}).get("username", 0)),
+                "categories": [class_json.get("category")],
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: No `[None]` should be possible.

File: youtube_dl/extractor/skillshare.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+        lessons_json = []
+        # Pretty sure all classes only have one unit but flattening just in case.
+        for unit_json in class_json["_embedded"]["units"]["_embedded"]["units"]:
+            lessons_json += (unit_json["_embedded"]["sessions"]["_embedded"]["sessions"])
+
+        videos = []
+        for lesson_json in lessons_json:
+            lesson_thumbnail_urls = [
+                lesson_json.get("video_thumbnail_url", ""),
```
Comment (dstftw) [inline youtube_dl/extractor/skillshare.py]: Processing empty strings is senseless.

File: youtube_dl/extractor/skillshare.py (hunk 2 above)
Comment (Oshawk) [reply youtube_dl/extractor/skillshare.py]: Should everything be single quotes?

File: youtube_dl/extractor/skillshare.py (hunk 4 above)
Comment (Oshawk) [reply youtube_dl/extractor/skillshare.py]: My thought on the base class was to make it simpler to allow more features such single video / playlist downloading later on. @dstftw

File: youtube_dl/extractor/skillshare.py (hunk 2 above)
Comment (Oshawk) [reply youtube_dl/extractor/skillshare.py]: Oh, you mean make double?

**Comment by cmp** [comment]
I have tested this pull request - works for paid and free content.
My only suggestion please add: user_type "trial"   
i.e.:
...
    elif user_type == "Premium Member" or "trial":
...

"Trial Membership" is essentialy a "Premium Membership"

**Comment by bdrtsky** [comment]
Is it working?

**Comment by Ruthalas** [comment]
Can this be merged?

**Comment by Oshawk** [comment]
It seems like this will never get merged. Closing.

**Pull request closed without merging**